src/coap.c

```diff
@@ -308,4 +308,182 @@ NS_INTERNAL uint32_t coap_parse(struct iobuf *io, struct ns_coap_message *cm) {
   return cm->flags;
 }
 
-#endif  /* NS_ENABLE_COAP */
+/* Calculates extended size of given Opt Number/Length in coap message. */
+static size_t coap_get_ext_opt_size(uint32_t value) {
+  int ret = 0;
+
+  if (value >= 13 && value <= 0xFF + 13) {
+    ret = sizeof(uint8_t);
+  } else if (value > 0xFF + 13 && value <= 0xFFFF + 269) {
+    ret = sizeof(uint16_t);
+  }
+
+  return ret;
+}
+
+/* Splits given Opt Number/Length into base and ext values. */
+static int coap_split_opt(uint32_t value, uint8_t *base, uint16_t *ext) {
+  int ret = 0;
+
+  if (value < 13) {
+    *base = value;
+  } else
+  if (value >= 13 && value <= 0xFF + 13) {
+    *base = 13;
+    *ext = value - 13;
+    ret = sizeof(uint8_t);
+  } else if (value > 0xFF + 13 && value <= 0xFFFF + 269) {
+    *base = 14;
+    *ext = value - 269;
+    ret = sizeof(uint16_t);
+  }
+
+  return ret;
+}
+
+/* Puts uint16_t (in network order) into given char stream. */
+static char *coap_add_uint16(char *ptr, uint16_t val) {
+  uint16_t msg_id = htons(val);
+  *ptr = msg_id & 0x00FF;
```

Comment:
here you're writing the LSB of msg_id into the first byte, and then the MSB in the next byte. This is little endian encoding. Since you previously swapped the bytes with `htons`, this effectively big endian encoding, but this works only on a little endian arch, since `htons` swaps bytes only on that platform.

The right code is simply:

```c
*ptr++ = val >> 8;
*ptr++ = val & 0xFF;
return ptr;
```